### ISSUE TITLE
Introduce type hints for static type checking

### ISSUE THREAD

**Issue by squiddy** [unknown-horizons/unknown-horizons#2513]
[PEP-484](https://www.python.org/dev/peps/pep-0484/) introduced new syntax to declare type hints. The interpreter makes no use of them, other then to expose them on e.g. function objects.

Third party tools like [mypy](http://mypy-lang.org/) understand the syntax and can check programs statically.

Note: this won't affect the game at runtime at all. It won't run faster, it won't run slower. It will however most likely help catching bugs and making lovers of static typing happy.

If I understand it correctly, mypy will only check function bodies if we start providing type hints, however running it against our codebase produces some errors. Failure to import FIFE and some complaints about our class inheritance trees. 

- [ ] Fix issues raised when running mypy on clean codebase
- [ ] Identify parts of codebase most likely to benefit the most (I'm thinking of essentials, Scheduler and stuff) and add type hints
- [ ] Profit?
- [ ] Nice to have: add mypy checks to CI

@ChrisOelmueller COMEFROM .idle

**Comment by MasterofJOKers** [comment]
I've read in an article about mypy, that one should start with getting '--check-untyped-defs' running and then continue to add explicit typing to the rest of the code, once this runs without errors and is added to the CI.

**Comment by squiddy** [comment]
Indeed. The article in question (I believe it's [the one from Tulip](http://blog.zulip.org/2016/10/13/static-types-in-python-oh-mypy/)), says to start off with `mypy --py2 --silent-imports --fast-parser -i <paths>`. Once that works, one should add `--check-untyped-defs` what you mentioned.

I've updated the issue accordingly.

**Comment by squiddy** [comment]
I'd say we skip checking the tests for now, especially in GUI tests a lot of monkey-patching is going on.

**Comment by squiddy** [comment]
I'm working on the first step and I'm wondering how we should handle the following problem (so far I haven't found clues in the documentation):

```python
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
	from horizons.util.uhdbaccessor import UhDbAccessor

# The global Fife instance. This is used both inside a game session and in the menus.
fife = None # type: Any

# The global read-only data database (includes most content/*.sql)
db = None # type: UhDbAccessor
```

This is `horizons/globals.py` with type hints. Even when using comments, one needs to import types. Running this would error since `typing` is not installed. An easy solution could be to ship `typing` with UH, but that doesn't sound right.

**Comment by MasterofJOKers** [comment]
If I understand it correctly, one can do something along the lines of `if False: import bla`. It's mentioned in the article you linked for python2 code. Although typing.TYPE_CHECKING seems to be the new way to do it.

**Comment by squiddy** [comment]
I thought that was merely about imports solely for the purpose of type hints, that if always imported, could cause import cycles. But you're right, the following should work:

```python
import my
import normal
import stuff

if False:
    from typing import Any
    from some.package.somewhere import Somethnig
```

**Comment by squiddy** [comment]
However, once we move to python3 this won't work I think, because the type hints become part of the code. Therefore they need to be imported unconditionally. I've read something about forward references, where types are written as strings. But that probably won't work for the basic types like `List`, `Dict`. I'll ask at mypy's gitter channel, see if they have any insights.

**Comment by MasterofJOKers** [comment]
Did mypy's gitter turn up with a solution?

**Comment by squiddy** [comment]
Oh noes. I completely forgot about it, because I was busy asking for help with contributing a bugfix. I can do that later today if no one else is quicker.

**Comment by MasterofJOKers** [comment]
"Including it in the package" sounds like `horizons.ext` to me.

**Comment by squiddy** [comment]
Yeah, I really don't want to do that. Which is why I suggested somehow including it in the packaging process. mypy is a development tool, I should be able to use the latest version (especially because libraries around mypy are still in flux), without regularly updating the file in `horizons.ext`.

But, for the time beeing, I suppose we could go with `if False`, focus on the work with mypy, and figure this out later. It's not really an issue right now, only in python 3.

**Comment by MasterofJOKers** [comment]
IIRC we don't need it at runtime, so the better way for packaging would be to patch all the imports out of the project. As you said, it's only for development.

**Comment by squiddy** [comment]
That won't work in Python 3 as I said. There you'd write

```python
def foo(x: Any) -> List[int]:
  return [1,2,3]
```

instead of

```python
def foo(x):
  # type: (Any) -> List[int]
  return [1,2,3]
```

which means you need Any & List defined at runtime. I guess we could patch those out as well, but I don't think this is something we actually should do.

**Comment by MasterofJOKers** [comment]
And python3 still hasn't typing included?

**Comment by squiddy** [comment]
I can't recall any plans to ever include it.

The PEP was written with no specific implementation in mind. You can define type hints, but the interpreter doesn't do anything with them (other than parsing and exposing the information e.g. on the func object). What to do with them is left to third-party packages.

AFAIK typing happens to be such a third-party package that is developed by (at least one) python core developer(s). It merely exposes some classes to define type hints, e.g. `List[int]`. That's why it's "only" 2000 lines long. The information is used by `mypy` (which is probably much bigger) to provide static type checks.

**Comment by MasterofJOKers** [comment]
```python
$ python3
Python 3.5.2+ (default, Sep 10 2016, 10:24:58) 
[GCC 6.2.0 20160901] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import typing
>>> 
```
It seems to be.

**Comment by squiddy** [comment]
Oh. Guess I was wrong. So the stuff `typing` does is then part of the PEP, I see. I'm a bit surprised, I thought they were actively developing typing as well.

So... sooner or later we will have `typing` imports unconditionally in the code then. Looking at typing's README:

> - May 24, 2015: Python 3.5.0 beta 1 -- PEP 484 accepted, typing.py feature complete and checked into CPython repo
> - August 9, 2015: Python 3.5.0 release candidate 1 -- Last chance for fixes to typing.py barring emergencies:
> - September 13, 2015: Python 3.5.0 final release

I suppose at this point the library is pretty solid (at least for our use-case), and we can include it for python2? Adding some sys.path modification to be able to directly import typing, no `horizons.ext.typing` stuff.

**Comment by squiddy** [comment]
I think this then is the way to go:

1. include python2 backport of typing in `horizons.ext`
2. adapt imports to read `from horizons.ext.typing import Any, ...` (unconditionally)
3. once we switched to python3, replace `from horizons.ext.typing` with `from typing`
4. remove `horizons.ext.typing`

Does that sound good @MasterofJOKers ?

**Comment by MasterofJOKers** [comment]
Sounds good to me.

**Comment by squiddy** [comment]
Packages like `horizons.gui.modules`, `horizons.gui.keylistener` and `horizons.gui.tabs` trigger an bug in mypy: https://github.com/python/mypy/issues/2052

We can work-around this by using relative imports there, however we haven't really decided if that's something we want to do (I'd prefer it).

**Comment by LinuxDonald** [comment]
I'm an code noob but does it makes code easier to ready?

**Comment by squiddy** [comment]
> I'm an code noob but does it makes code easier to ready?

The advantage is that one can restructure whole modules (like `horizons.gui.tabs`) without renaming imports of files inside that directory. Which will rarely happen in our case though.

It's probably a matter of taste. I do find it superfluous to repeat the module path, e.g. in `horizons.gui.tabs` we do:

```python
from horizons.gui.tabs.foo import FooTab
```

where we could write

```python
from .foo import FooTab
```

instead.

However, I'm fine with both styles. It's just that the first one is causing problems for type checking right now.

**Comment by LinuxDonald** [comment]
I think the second style is better because.

(Weil man sich nicht mehr so sehr verschreiben kann)

**Comment by squiddy** [comment]
Step one is *almost* done. Ignoring errors from global gettext functions we get this: https://gist.github.com/squiddy/c6dd6479da8cd5b7495374fd4ba85acf

The majority of errors is of the form:
`Definition of "<A>" in base class "<B>" is incompatible with definition in base class "<C>"`

**Comment by squiddy** [comment]
All that's left are those incompatible errors. The one in production overview to me is not all clear whether that should be an error or not:

> Definition of "show" in base class "StatsWidget" is incompatible with definition in base class "Window"

I've opened a new issue at mypy https://github.com/python/mypy/issues/2377.

The rest of the errors might actually be somewhat related. As far as I understand the code, we aren't doing super calls inside `__init` for example, since that's called done in `__init__` and `load`. In our case, mypy shouldn't complain, but I can see why it does, since we're using the same name everywhere.

**Comment by squiddy** [comment]
First step (`Fix issues raised when running mypy on clean codebase (mypy --py2 --silent-imports --fast-parser -p horizons)`) is done. We still get lots of issues related to inheritance (https://github.com/unknown-horizons/unknown-horizons/issues/2513#issuecomment-257078652), but that's nothing we can fix or ignore with mypy right now. I think we probably write some `run_mypy_check` script that ignores those files completely for now.

Next step is: `Fix issues raised by adding the --check-untyped-defs argument`

**Comment by AndyMender** [comment]
Not sure how far this has gone, but I read recently in the Python 3.6 Release Notes that static typing is an optional built-in feature now. However, Ubuntu still uses 3.4 (Trusty) or 3.5 (Zesty) as the default Python3 version so we might want to wait a little longer :).

**Comment by squiddy** [comment]
Variable annotations (Not Übung comments) are now possible. Other than
that, we've already been using this. This thread and commits are proof of
that. There ist no point in waiting.

Am 29.04.2017 10:28 nachm. schrieb "Andrzej Bylicki" <
notifications@github.com>:

**Comment by squiddy** [comment]
So apparantly we planned to switch over to python 3.5 when we started working on this. The typing module was only included in python3.5, that's why isort is complaining now because it detects typing as a third-party module. Will fix it now.

**Comment by LinuxDonald** [comment]
@squiddy any news about this?
